File: python/openharness/core/events.py
"""Streaming event types for the agent loop."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union


@dataclass(frozen=True)
class TextDelta:
    """A chunk of streamed text from the LLM."""

    content: str


@dataclass(frozen=True)
class ToolCallStart:
    """LLM has requested a tool call."""

    tool_name: str
    call_id: str


@dataclass(frozen=True)
class ToolCallEnd:
    """A tool call has finished executing."""

    call_id: str
    output: str
    is_error: bool = False


@dataclass(frozen=True)
class PermissionRequest:
    """Agent is asking for user permission to run a tool."""

    tool_name: str
    call_id: str
    arguments: dict
    risk_level: str


@dataclass(frozen=True)
class CostUpdate:
    """Token usage and cost for an API call."""

    input_tokens: int
    output_tokens: int
    cost: float
    model: str


@dataclass(frozen=True)
class TurnComplete:
    """The agent has finished its turn (no more tool calls)."""

    reason: str  # "completed", "aborted", "max_turns", "error"


@dataclass(frozen=True)
class ErrorEvent:
    """An error occurred during the agent loop."""

    message: str
    recoverable: bool = False


# Union type for all events that can be yielded by the agent loop
Event = Union[
    TextDelta,
    ToolCallStart,
    ToolCallEnd,
    PermissionRequest,
    CostUpdate,
    TurnComplete,
    ErrorEvent,
]

File: src/types/events.ts
/**
 * Streaming event types for the agent loop.
 */

export type TextDelta = {
  readonly type: "text_delta";
  readonly content: string;
};

export type ToolCallStart = {
  readonly type: "tool_call_start";
  readonly toolName: string;
  readonly callId: string;
};

export type ToolCallComplete = {
  readonly type: "tool_call_complete";
  readonly callId: string;
  readonly toolName: string;
  readonly arguments: Record<string, unknown>;
};

export type ToolCallEnd = {
  readonly type: "tool_call_end";
  readonly callId: string;
  readonly output: string;
  readonly isError: boolean;
};

export type PermissionRequest = {
  readonly type: "permission_request";
  readonly toolName: string;
  readonly callId: string;
  readonly description: string;
  readonly riskLevel: string;
};

export type CostUpdate = {
  readonly type: "cost_update";
  readonly inputTokens: number;
  readonly outputTokens: number;
  readonly cost: number;
  readonly model: string;
};

export type TurnComplete = {
  readonly type: "turn_complete";
  readonly reason: string;
  readonly sessionId?: string;
};

export type ErrorEvent = {
  readonly type: "error";
  readonly message: string;
};

export type RateLimited = {
  readonly type: "rate_limited";
  readonly retryIn: number;   // seconds
  readonly attempt: number;   // 1-based
};

export type ThinkingDelta = {
  readonly type: "thinking_delta";
  readonly content: string;
};

export type ToolOutputDelta = {
  readonly type: "tool_output_delta";
  readonly callId: string;
  readonly chunk: string;
};

export type AskUserRequest = {
  readonly type: "ask_user";
  readonly callId: string;
  readonly question: string;
  readonly options?: readonly string[];
};

export type StreamEvent =
  | TextDelta
  | ToolCallStart
  | ToolCallComplete
  | ToolCallEnd
  | ToolOutputDelta
  | PermissionRequest
  | AskUserRequest
  | CostUpdate
  | TurnComplete
  | ErrorEvent
  | RateLimited
  | ThinkingDelta;
